In [1]:

#importing
import tensorflow as tf
from tensorflow import keras
import pandas as pd

from keras.models import Sequential

In [11]:
#About Dataset

# Each record consists of three attributes:

# is_sarcastic: 1 if the record is sarcastic otherwise 0
# headline: the headline of the news article
# article_link: link to the original news article. Useful in collecting supplementary data

# from google.colab import drive
# drive.mount('/content/drive')

Mounted at /content/drive


In [13]:
import json


import json
datastore = [json.loads(line) for line in open("/content/sample_data/Sarcasm_Headlines_Dataset_v2.json", 'r')]    #need to upload this file everytime with new runtime
  
sentences = []
labels = []
urls = []

for item in datastore:
  sentences.append(item["headline"])
  labels.append(item["is_sarcastic"])
  urls.append(item["article_link"])

In [31]:
datastore[0:5]

[{'article_link': 'https://www.theonion.com/thirtysomething-scientists-unveil-doomsday-clock-of-hai-1819586205',
  'headline': 'thirtysomething scientists unveil doomsday clock of hair loss',
  'is_sarcastic': 1},
 {'article_link': 'https://www.huffingtonpost.com/entry/donna-edwards-inequality_us_57455f7fe4b055bb1170b207',
  'headline': 'dem rep. totally nails why congress is falling short on gender, racial equality',
  'is_sarcastic': 0},
 {'article_link': 'https://www.huffingtonpost.com/entry/eat-your-veggies-9-delici_b_8899742.html',
  'headline': 'eat your veggies: 9 deliciously different recipes',
  'is_sarcastic': 0},
 {'article_link': 'https://local.theonion.com/inclement-weather-prevents-liar-from-getting-to-work-1819576031',
  'headline': 'inclement weather prevents liar from getting to work',
  'is_sarcastic': 1},
 {'article_link': 'https://www.theonion.com/mother-comes-pretty-close-to-using-word-streaming-cor-1819575546',
  'headline': "mother comes pretty close to using wor

In [29]:
urls[0]

'https://www.theonion.com/thirtysomething-scientists-unveil-doomsday-clock-of-hai-1819586205'

In [30]:
labels[0]

1

In [35]:
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences

tokenizer = Tokenizer(oov_token="<OOV>")   #alloting tokens to words
#asking tokenizer to go to text and fix them in sentences     
tokenizer.fit_on_texts(sentences)

word_index = tokenizer.word_index
print("word index",word_index)
print("length of word index",len(word_index))

sequences = tokenizer.texts_to_sequences(sentences)
print("sequences",sequences)
padded = pad_sequences(sequences, padding="post")

print(sentences[2])
print(padded[2])

word index {'<OOV>': 1, 'to': 2, 'of': 3, 'the': 4, 'in': 5, 'for': 6, 'a': 7, 'on': 8, 'and': 9, 'with': 10, 'is': 11, 'new': 12, 'trump': 13, 'man': 14, 'at': 15, 'from': 16, 'about': 17, 'by': 18, 'after': 19, 'you': 20, 'this': 21, 'out': 22, 'up': 23, 'be': 24, 'as': 25, 'that': 26, 'it': 27, 'how': 28, 'not': 29, 'he': 30, 'his': 31, 'are': 32, 'your': 33, 'just': 34, 'what': 35, 'all': 36, 'who': 37, 'has': 38, 'will': 39, 'report': 40, 'into': 41, 'more': 42, 'one': 43, 'have': 44, 'year': 45, 'over': 46, 'why': 47, 'day': 48, 'u': 49, 'area': 50, 'woman': 51, 'can': 52, 's': 53, 'says': 54, 'donald': 55, 'time': 56, 'first': 57, 'like': 58, 'no': 59, 'her': 60, 'get': 61, 'off': 62, 'old': 63, "trump's": 64, 'life': 65, 'now': 66, 'people': 67, "'": 68, 'an': 69, 'house': 70, 'still': 71, 'obama': 72, 'white': 73, 'back': 74, 'make': 75, 'was': 76, 'than': 77, 'women': 78, 'if': 79, 'down': 80, 'when': 81, 'i': 82, 'my': 83, '5': 84, 'clinton': 85, 'could': 86, 'they': 87, 'th

In [19]:
print(padded[0])

[16004   355  3167  7474  2644     3   661  1119     0     0     0     0
     0     0     0     0     0     0     0     0     0     0     0     0
     0     0     0     0     0     0     0     0     0     0     0     0
     0     0     0     0     0     0     0     0     0     0     0     0
     0     0     0     0     0     0     0     0     0     0     0     0
     0     0     0     0     0     0     0     0     0     0     0     0
     0     0     0     0     0     0     0     0     0     0     0     0
     0     0     0     0     0     0     0     0     0     0     0     0
     0     0     0     0     0     0     0     0     0     0     0     0
     0     0     0     0     0     0     0     0     0     0     0     0
     0     0     0     0     0     0     0     0     0     0     0     0
     0     0     0     0     0     0     0     0     0     0     0     0
     0     0     0     0     0     0     0     0]


We can see that it has been encoded with the values for the keys that are the corresponding word in the sentence. This is the size of the padded matrix. 

In [20]:
print(padded.shape)

(28619, 152)


We had 28,619 sentences, and they were encoded with padding, to get them up to 152 words long which was the length of the longest one. You could truncate this if you like, but I'll keep it at 152. That's it for processing the Sarcasm data set.

In [22]:
print(padded[10])

[  21 2645   11 3360    7  426    5 3580    9   69 1006   43   15   26
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0]
